src/lib/operators/join_mpsm.cpp

```diff
@@ -0,0 +1,498 @@
+      auto right_rest = TableRange(right_node_id, right_cluster_id, right_run_start, right_size);
+      if (right_run_start < right_size) {
+        _join_runs(left_rest, right_rest, CompareResult::Greater, left_joined);
+      }
+    }
+
+    if (_mode == JoinMode::Left || _mode == JoinMode::Outer) {
+      _emit_right_null_combinations(left_node_id, left_cluster_id, left_cluster, left_joined);
+    }
+  }
+
+  /**
+  * Performs the join on all clusters in parallel.
+  **/
+  void _perform_join() {
+    std::vector<std::shared_ptr<AbstractTask>> jobs;
+
+    // Parallel join for each cluster
+    // TODO(florian): this is ideal, just make sure the jobs get scheduled on the correct nodes
+    for (size_t cluster_number = 0; cluster_number < _cluster_count; ++cluster_number) {
+      jobs.push_back(std::make_shared<JobTask>([this, cluster_number] { this->_join_cluster(cluster_number); }));
+      jobs.back()->schedule();
+    }
+
+    CurrentScheduler::wait_for_tasks(jobs);
+  }
+
+  /**
+  * Concatenates a vector of pos lists into a single new pos list.
+  **/
+  std::shared_ptr<PosList> _concatenate_pos_lists(std::vector<std::vector<std::shared_ptr<PosList>>>& pos_lists) {
+    auto output = std::make_shared<PosList>();
+
+    // Determine the required space
+    size_t total_size = 0;
+    for (const auto& partition_lists : pos_lists) {
+      for (const auto& pos_list : partition_lists) {
+        total_size += pos_list->size();
+      }
+    }
+
+    // Move the entries over the output pos list
+    output->reserve(total_size);
+    for (const auto& partition_lists : pos_lists) {
+      for (const auto& pos_list : partition_lists) {
+        output->insert(output->end(), pos_list->begin(), pos_list->end());
+      }
+    }
+
+    return output;
+  }
+
+  /**
+  * Adds the columns from an input table to the output table
+  **/
+  void _add_output_columns(std::shared_ptr<Table> output_table, std::shared_ptr<const Table> input_table,
+                           std::shared_ptr<const PosList> pos_list) {
+    auto column_count = input_table->column_count();
+    for (ColumnID column_id{0}; column_id < column_count; ++column_id) {
+      // Add the column definition
+      auto column_name = input_table->column_name(column_id);
+      auto column_type = input_table->column_type(column_id);
+      output_table->add_column_definition(column_name, column_type);
+
+      // Add the column data (in the form of a poslist)
+      // Check whether the referenced column is already a reference column
+      const auto base_column = input_table->get_chunk(ChunkID{0})->get_column(column_id);
+      const auto ref_column = std::dynamic_pointer_cast<const ReferenceColumn>(base_column);
+      if (ref_column) {
+        // Create a pos_list referencing the original column instead of the reference column
+        auto new_pos_list = _dereference_pos_list(input_table, column_id, pos_list);
```

Comment:
It does not use any member values, so definetly a free function. Where should we best put it @Bouncner ?

Reply:
Good question. @mrzzzrm: is this also related to your recent hash join changes?

Reply:
looks like this join operator could benefit from something similar to what I did to the JoinHash: You're redundantly re-referencing pos lists, since it is very likely that neighbouring columns in the input table share the same pos lists.